rfcs/20230420-kernel-cache/README.md

```diff
@@ -0,0 +1,292 @@
+    value_impl_t() = default;
+    virtual ~value_impl_t() = default;
+
+    value_impl_t(const value_impl_t &) = delete;
+    value_impl_t &operator=(const value_impl_t &) = delete;
+};
+} // kernel_cache
+```
+
+#### Example
+
+Below is a pseudo-code that demonstrates the idea of using the aforementioned
+abstractions for GPU engine kind.
+```cpp
+// GPU specific `key` and `value` implementations.
+struct gpu_key_t : public key_impl_t {
+    // This could be a serialized configuration structure or anything else.
+    std::vector <uint8_t> data;
+
+    compare(const key_impl_t *key_impl) const override {
+        const auto *typed_key = static_cast<const gpu_key_impl_t *>(key_impl);
+        return data == typed_key->data;
+    }
+
+    size_t hash() const override {
+        // This is just for demonstration purposes. In the implementation
+        // we should use `hash_combine` function.
+        size_t seed = 9;
+        for (uint8_t val : data)
+            seed ^= (val << 1);
+        return seed; ^
+    }
+};
+
+struct gpu_value_t : public value_impl_t {
+    compute::kernel_t kernel;
+};
+
+`struct kernel_cache_t {
+    /* Internal interfaces */
+    /* ... */
+    lru_cache_t<key_t, value_t, value_t> cache;
+};
+
+// Returns a reference to a kernel cache initialized with the
+// primitive cache capacity.
+kernel_cache_t &global_kernel_cache();
+
+// Inside sycl_gpu_engine_t::create_kernel(...)
+auto &kcache = kernel_cache::global_kernel_cache();
+// Create a key.
+kernel_cache::key_t key = ...;
+kernel_cache::value_t value = kcache.get(key);
+
+// Assume bool(value) == true.
+gpu_value_t *gpu_value = static_cast<gpu_value_t *>(value.impl().get());
+compute::kernel_t k = gpu_value->kernel;
+// Now we can use the kernel `k`.
+```
+
+#### Memory Consumption Management
+
+oneDNN users usually want to have control over amount of memory that the library
+consumes if that amount is significant. Introducing the cache will come with
+additional memory consumption though if it is reasonably small then users may not
+want to deal with it. Given that, the are two approaches to the cache memory
+consumption.
+
+##### Option 1: Keep the Cache Hidden (recommended)
+
+This option suggests making the cache a hidden feature without providing any
+control over its capacity.
+
+There are two approaches to defining the cache capacity:
+1. Define a reasonable constant
+2. Tie the cache capacity to the primitive cache capacity
+
+The first approach is overly rigid and it is hard to define a reasonable
+capacity that would work for the majority of workloads therefore we will consider
+the second one.
+
+Pros:
+* Provides sufficient control to benefit from the cache and to maintain memory
+consumption
+* Doesn't introduce yet another *cache capacity* definition that will simplify using the cache
+
+Cons:
+* There is no guarantee that all kernels in the kernel cache are used by all the
+primitives in the primitive cache. It might happen those primitives in the
```

Comment:
Depending on how the kernel cache is tied to the primitive cache, this might not be an issue. For example, if only primitives in the primitive cache get their kernels cached, and eviction from the primitive cache reduces the reference count of kernels in the kernel cache, then there should be no extra kernels.

Reply:
> Depending on how the kernel cache is tied to the primitive cache, this might not be an issue. For example, if only primitives in the primitive cache get their kernels cached, and eviction from the primitive cache reduces the reference count of kernels in the kernel cache, then there should be no extra kernels.

Is this how life cycle of the kernels in the kernel cache managed? If so, there is indeed no need to allow the control of "kernel cache" sizes. The option controlling primitive cache sizes controls kernel cache sizes indirectly. The size of "kernel cache" won't exceed that of "primitive cache".

Reply:
The primitive and kernel caches are not coherent in that aspect unless we implement a mechanism to make them coherent. But that would be a very complex implementation with a high matinanace cost.

Suppose each primitive has only 1 corresponding kernel. Then we fill the primitive cache up to its capacity.
At this point the kernel cache size can be less or equal to the primitive cache size because some kernels can be re-used by the primitives in the primitive cache. So we indeed don't consume extra memory here.
 
The discrepancy appears when the size of the kernel cache reaches its capacity and the eviction mechanism kicks in. For example, we want to add one more primitive that is not in the primitive cache and that has a kernel that hasn't been added to the kernel cache yet. It means we need to evict a primitive from the primitive cache. Since least frequently used primitive doesn't nesasariliy correspond to the least frequently used kernel it might happen that the primitive is getting evicted from the primitive cache but its kernel stays in the kernel cache. Instead of the kernel, another kernel gets evicted even though the primitive in the primitive cache, which uses the evicted kernel, still holds a reference to it, therefore the memory is not released. So we end up having `"kernel cache capacity" + 1` number of kernels. There can be more extra kernels because of that.

Given that the size of a kernel is approximately between ~30kb and ~200kb for GPU and up to ~500kb for CPU we can assume that overall the extra memory consumption will not be a problem. And if it becomes a problem we can add an API and an environment variable to control kernel cache independently from the primitive cache.

Reply:
I see. So the added complexity of tightly coupled caches is not worth the extra development and maintenance effort. That is fine, but I think the proposed coupling mechanism should be more explicitly stated in the RFC. From this discussion it sounds like the capacity of the kernel cache will be the same as the capacity of the primitive cache (at least to begin with), but in the RFC it just states "Tie the cache capacity to the primitive cache capacity" which is a little vague.

Reply:
@densamoilov I think this is a good idea to have an LRU policy on the kernel cache, but quick questions here:
- when will that data be updated? Is it only upon primitive creation, or would a primitive cache hit also update the kernel cache statistics?
- Many implementations rely on multiple kernels (e.g. brgemm on CPU, matmul on GPU), so having `kernel_cache_capacity = primitive_cache_capacity` seems risky. Did you gather some data on that?